### PR TITLE
[final] add should purchase promo product

### PR DESCRIPTION
**Pull request by aboedo**
### Description
https://trello.com/c/emjo2Nlf

Related PR: https://github.com/RevenueCat/react-native-purchases/pull/98

~This is still missing cleanup work, but it's functional.~

### Changes summary: 
- bumped iOS sdk to v3.0.2, updated path to `Purchases.framework` in script
- added method `makeDeferredPurchase`, to be called by client apps when they're ready to make a promotional purchase. 
- added method `shouldPurchasePromoProduct`, called by the SDK when it detects that a promotional purchase should be made. 
- added `addShouldPurchasePromoProductListener` and `removeShouldPurchasePromoProductListener`, to be called by client apps so they can set up listeners. They will get a function that they call to trigger the purchase on our side, which they can call whenever the app is ready to make the purchase. 

### Code flow: 

- client calls set up purchases. 
- this sets up a callback id, to be used later when promo purchases are detected. 
- client registers listeners for when promo purchases are ready to be made. listeners are functions, and they should internally call `makeDeferredPurchase();`. listeners won't be executed until promo purchases are detected. 
- we store the listeners (clients can remove them if needed). 
- app is used, time passes
- promo purchase is detected by app
- sdk calls `shouldPurchasePromoProduct`. we store a callback id to identify the exact call. 
- sdk notifies client that it's ready to make a deferred purchase. Listeners are notified, they execute client code and call `makeDeferredPurchase()`
- `makeDeferredPurchase` gets the callback id to identify the exact call, and calls `makeDeferredPurchase` on the sdk.

### CODE REVIEW

File: src/plugin/plugin.ts [hunk 1]
```diff
@@ -843,6 +848,55 @@ class Purchases {
...
+   * @param {ShouldPurchasePromoProductListener} shouldPurchasePromoProductListener Called when a user initiates a
+   * promotional in-app purchase from the App Store. If your app is able to handle a purchase at the current time, run
+   * the deferredPurchase function. If the app is not in a state to make a purchase: cache the deferredPurchase, then
+   * call the deferredPurchase when the app is ready to make the promotional purchase.
+   * If the purchase should never be made, you don't need to ever call the deferredPurchase and the app will not
+   * proceed with promotional purchases.
+   */
+  public static addShouldPurchasePromoProductListener(shouldPurchasePromoProductListener: ShouldPurchasePromoProductListener): void { 
+    if (typeof shouldPurchasePromoProductListener !== "function") {
+      throw new Error("addPurchaserInfoUpdateListener needs a function");
```
Comment (vegaro) [inline src/plugin/plugin.ts]: The error should say "addShouldPurchasePromoProductListener needs a function"

Comment (aboedo) [inline src/plugin/plugin.ts]: Oops, fixing

File: src/plugin/plugin.ts [hunk 2]
```diff
@@ -843,6 +848,55 @@ class Purchases {
...
+    if (shouldPurchasePromoProductListeners.indexOf(listenerToRemove) !== -1) {
+      shouldPurchasePromoProductListeners = shouldPurchasePromoProductListeners.filter(
+        listener => listenerToRemove !== listener
+      );
+      return true;
+    }
+    return false;
+  }
+
+  private static setupShouldPurchasePromoProductCallback() {
```
[the remark this answers is not in the document]
Comment (vegaro) [reply src/plugin/plugin.ts]: I love it

File: src/plugin/plugin.ts (hunk 1 above)
Comment (vegaro) [reply src/plugin/plugin.ts]: Cool. I commented on my review on this

File: src/plugin/plugin.ts [hunk 3]
```diff
@@ -843,6 +848,55 @@ class Purchases {
...
+      );
+      return true;
+    }
+    return false;
+  }
+
+  private static setupShouldPurchasePromoProductCallback() { 
+    window.cordova.exec(
+      (callbackResult: any) => {
+          const callbackID = callbackResult.callbackID;
```
Comment (vegaro) [inline src/plugin/plugin.ts]: This two lines could be removed with:
`({callbackID}:{callbackID: number}) => {`

File: src/plugin/plugin.ts (hunk 1 above)
Comment (aboedo) [reply src/plugin/plugin.ts]: @vegaro https://github.com/RevenueCat/react-native-purchases/pull/98/files#diff-f41e9d04a45c83f3b6f6e630f10117feR546 LOL 😂 copy/paste kings over here. I'll make a PR for that one as well.

File: src/plugin/plugin.ts (hunk 3 above)
Comment (aboedo) [reply src/plugin/plugin.ts]: cool! didn't know that! I'll update

File: src/plugin/plugin.ts [hunk 4]
```diff
@@ -843,6 +848,55 @@ class Purchases {
...
+      shouldPurchasePromoProductListeners = shouldPurchasePromoProductListeners.filter(
+        listener => listenerToRemove !== listener
+      );
+      return true;
+    }
+    return false;
+  }
+
+  private static setupShouldPurchasePromoProductCallback() { 
+    window.cordova.exec(
```
[the remark this answers is not in the document]
Comment (aboedo) [reply src/plugin/plugin.ts]: good catch. I'll fix it and update. I forgot to test on android (hides in shame)

File: src/plugin/plugin.ts (hunk 3 above)
Comment (aboedo) [reply src/plugin/plugin.ts]: it'd be `({callbackResult}:{callbackID: number}) = {`, right?

**Comment by aboedo** [comment]
I'm not sure what's going on with `src/ios/Purchases.framework.Purchases` on github. I see it locally, and it was modified on this commit: 
https://github.com/revenuecat/cordova-plugin-purchases/commits/438fc31d47d7933b53b9c81567aa3ad7f520ea1e
however, `git show 438fc31d47d7933b53b9c81567aa3ad7f520ea1e` shows it as modified
```
diff --git a/src/ios/Purchases.framework/Purchases b/src/ios/Purchases.framework/Purchases
index 7899882..e41cb3f 100755
Binary files a/src/ios/Purchases.framework/Purchases and b/src/ios/Purchases.framework/Purchases differ
```

I tried clean-cloning and checking out the branch and I do get the file, so I'm not sure what's going on. I'll do some further digging, but right now it seems that Github is the only one that sees the file as removed.

**Comment by aboedo** [comment]
you can even see the file on the branch: https://github.com/RevenueCat/cordova-plugin-purchases/blob/feature/addShouldPurchasePromoProduct/src/ios/Purchases.framework/Purchases

So I really think this is a github bug. If I try removing the file on my local repo and check git diff, it does show that I'm removing it.

**Comment by aboedo** [comment]
@vegaro would you mind doing a sanity check by checking out the branch?

File: src/android/PurchasesPlugin.java [hunk 5]
```diff
@@ -156,6 +156,11 @@ private void setAutomaticAppleSearchAdsAttributionCollection(boolean enabled, Ca
         // NOOP
     }
 
+    @PluginAction(thread = ExecutionThread.WORKER, actionName = "setupShouldPurchasePromoProductCallback")
+    private void setupShouldPurchasePromoProductCallback(CallbackContext callbackContext) {
+        // NOOP
```
Comment (aboedo) [inline src/android/PurchasesPlugin.java]: wild violation of Interface Segregation Principle, but I think it's okay? I copied it from line 156. 
other ideas would be to have the javascript side decide (which isn't ideal), or to somehow split interfaces, which might be too much work for too little gain.

**Comment by aboedo** [comment]
@vegaro ready for another pass

File: src/android/PurchasesPlugin.java (hunk 5 above)
Comment (vegaro) [reply src/android/PurchasesPlugin.java]: I think this is better too

**Pull request merged**